Add Filters Feature

Sabihashaik opened the pull request:
# Description
Current Filter Options are Black and White and Sepia. Current Layout of Filter Screen is very Basic.
User can save current Image and move forward to apply filter for next image. 
When user is done, he can click Done from the menubar. 
User can then click Create PDF to create a PDF from the filtered images. 
All filtered images are separately saved in a folder called PDFFilters on the User's SD Card or Internal Storage.

![filter_basic](https://user-images.githubusercontent.com/21033484/42412119-6c2eb248-8217-11e8-8b31-15153bc901ea.png)


I also noticed that this PR shows that it contains commits from the previous issue, I tried cherry Picking the commits to a different branch but all of them still show up in the PR. 
On Checking the code changes however, only the changes made in Add Filters Feature Show up.
Fixes #42 

## Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?
- [x] `./gradlew assembleDebug assembleRelease`
- [x] `./gradlew checkstyle`

# Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings

Swati4star commented:
Do we have only grayscale & sepia filter?

Sabihashaik commented:
@Swati4star As of now I have only added GrayScale and Sepia as I noticed you had mentioned that previously in the comments of the issue.
But if you are asking if the library has more filters, yes, it has access to all the filters from the [Effect Factory](https://developer.android.com/reference/android/media/effect/EffectFactory) and ability to [modify the Effect Factory Filters](https://github.com/burhanrashid52/PhotoEditor/wiki/Filter-Effect)

Swati4star commented:
oh okay. I don't think adding new filters will be that much effort later on. (will create further issues if we want more filters)

Let there be only these filters for now.
Amazing work! I'll just check it once locally.

Swati4star commented:
The options length is coming up to be uneven. Make it in one line.

<img width="327" alt="screen shot 2018-07-07 at 8 36 19 pm" src="https://user-images.githubusercontent.com/9164477/42412213-6f59e196-8225-11e8-8209-2f77bfe8849e.png">

Sabihashaik commented:
@Swati4star The Card view side depends on its contents I believe?  Since the Compress the Image has more text it's expanding.  Should I add more text to the "Filter Images" to make it even? Something like "Filter Images (GrayScale & Sepia)"

Swati4star commented:
Some problems user will face : 
- If I click on the right arrow button on filter images screen, you should automatically save it. (It's quite confusing. I am not editing any image & just clicking on right arrow button, & when I click on `Create PDF`, it shows `no images selected`

- See the below 2 screens, first one is real image & second one is after I applied effect to it:
<img width="329" alt="screen shot 2018-07-07 at 8 41 39 pm" src="https://user-images.githubusercontent.com/9164477/42412258-4ced57a4-8226-11e8-9bdb-df8b2f16488f.png">
<img width="363" alt="screen shot 2018-07-07 at 8 41 47 pm" src="https://user-images.githubusercontent.com/9164477/42412261-5d40721c-8226-11e8-8250-ceb65b1a24ca.png">

- Show the count below like `Showing 1/2 image`

- grayscale & Sepia buttons are not aligned.

- If user apply some effect & don't click save current, show an error message/alert like `The effects you have applied on this image are not saved yet. Do you want to save them?`.

Swati4star commented:
@Sabihashaik Just an extra new line at the end of the filter images method. That might solve the problem.

Sabihashaik commented:
@Swati4star Sure, will make the above mentioned changes.

Swati4star commented:
@Sabihashaik Add this library to here also : https://github.com/Swati4star/Images-to-PDF#dependencies

Sabihashaik commented:
@Swati4star Removing the save current and making the next button save automatically was generating errors and was not giving the desired output for the Created PDF File. Instead I Added Toasts where ever necessary.
Such as the user cannot proceed without clicking "Save Current" or Applying a Filter. 
I added the _Showing Image number_ feature.

Swati4star commented on app/src/main/java/swati4star/createpdf/fragment/HomeFragment.java:
```diff
@@ -283,8 +330,79 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
+            }
         }
     }
+    /**
```
I have added the morph button related functions in https://github.com/Swati4star/Images-to-PDF/blob/master/app/src/main/java/swati4star/createpdf/util/MorphButtonUtility.java file. There is no need to add these here.

Swati4star commented on app/src/main/java/swati4star/createpdf/activity/PhotoEditor.java:
```diff
@@ -0,0 +1,204 @@
+            e.printStackTrace();
+        }
+    }
+    private void grayscaleFilter() {
```
add comments to newly added function.

Sabihashaik commented:
@Swati4star Ok, I have added a None Filter Option.  I have added comments & Removed the morphing button. I modified the screen error. Please verify if the Screen error still shows.

Swati4star commented:
@Sabihashaik Few errors:
- The portrait image error is still there
- I selected 3 images. Clicked on `Filter Images`. and then, clicked on `Done`. Now, when I click on createPDF, it shows `No Images Selected`. This is kind of a major issue too. Look into this.

Sabihashaik commented:
@Swati4star The portrait image error, I noticed is being caused because I'm resizing the image to fit into the tiny square, so that the buttons aren't blocked.  It the image ratio which the user is using is not too big then the black background doesn't show. 
One way to fix this is to make the ImageView the Entire Screen Size and Shift the buttons to another fragment and such, I believe. What do you suggest?
I fixed the "Done" error.

Swati4star commented on app/src/main/res/layout/activity_photo_editor.xml:
```diff
@@ -0,0 +1,119 @@
+    <ja.burhanrashid52.photoeditor.PhotoEditorView
+        android:id="@+id/photoEditorView"
+        android:layout_width="wrap_content"
+        android:layout_height="350dp"
```
what if you set both to `wrap_content1`

Swati4star commented:
@Sabihashaik I tried by making PhotoEditorView's height wrap_content & added the whole layout in scrollview. It's working fine with that.

It's just we need to scroll down to add filter to image.
You can use frame layout & add a strip at bottom of screen with these options (None, grayscale & sepia), so that it is always visible on the screen, and only the PhotoEditorView is scrollable.

Swati4star commented:
One more issue,
I selected few images. clicked on filter images. I did nothing. I clicked on arrow, it says `Save current first` and when I select Save current, it says `Choose a filter first`. Don't make choosing filter compulsory, or don't show this message when no filter is added.

Swati4star commented:
Also, add back button at the top of title bar.  Use this code : https://stackoverflow.com/a/24422264/3623128

Sabihashaik commented:
@Swati4star Ok 👍 
Added:
1.  Back Button
2. Scroll View + Frame Layout
3. Removed Filter Toast.

Please let me know if there are any other changes I have to make

Swati4star commented:
@Sabihashaik I'll check it once locally.

Swati4star commented:
There is only one bug now:
- Select few images -> Filter images. Save first image. Click done. PDF has only 1 image.

Sabihashaik commented:
@Swati4star I have added a fix for the last bug.
